specify `if let` guards with updated scoping rules

dianne opened the pull request:
This is based on rust-lang/reference#1823 by @Kivooeo and @ehuss, rebased to resolve conflicts. Per https://github.com/rust-lang/rust/pull/141295#issuecomment-3172434801, I'm opening this as a separate PR to contribute an updated specification of `if let` guards' drop-scoping rules. The new spec is based on the compiler's implementation[^1], accounting for the changes in rust-lang/rust#143376 and additional corner-cases in examples. I've also done some minor edits to the sections on lexical scope and match expressions, but for the most part they're the same as in rust-lang/reference#1823.

[^1]: As with the rules for other syntactic constructs (e.g. `let` statements and `if` expressions), this is slightly simplified. Due to implementation details, the compiler's notion of drop scopes is more fine-grained than is necessary for the language spec.

Tracking issue: rust-lang/rust#51114
Stabilization: https://github.com/rust-lang/rust/pull/141295

Kivooeo committed:
added documentation for if let

Kivooeo committed:
cutted down match-expr, added info in destructors, fixed syntax, removed ingore from some blocks

ehuss committed:
Rework match guard chain

This applies some editorial rework, mainly to match the style of `if`
conditions.

dianne committed:
rework scoping rules for match arms and guards

dianne committed:
editorial tweaks

- Removes specification of `if let` guards' bindings "only" being valid
if the guard succeeds. The arm body is only executed if the guard
succeeds, so the bindings are always usable within the arm body when
it's executed.

- Makes terminology slightly more consistent (replaces some remaining
uses of "`if let` guard" with "`let` pattern")

- Other small wording and punctuation tweaks.

dianne commented on src/destructors.md:
```diff
@@ -255,9 +289,8 @@ smallest scope that contains the expression and is one of the following:
-  or a `match` guard.
-* The body expression for a match arm.
+* The non-pattern matching condition expression of an `if` or `while` expression or a non-pattern-matching `match` guard condition operand.
+* The pattern-matching guard, if present, and body expression for a `match` arm.
```
To address https://github.com/rust-lang/reference/pull/1823#issuecomment-3173098920 (cc @theemathas), the temporary lifetime of an `if let` guard scrutinee is always extended, like for `if let` expressions. There aren't lifetime extension rules like there are for `let` statements. This rule should hopefully cover that (similar to how the analogous rule for "The pattern-matching condition(s) and consequent body of `if`" does below). There's an example further down too showing that the scrutinee lives until the end of the arm. Do you think this is sufficient, or would this benefit from additional clarification?

theemathas replied:
This code currently compiles. I assumed that it does so due to temporary lifetime extension. Without lifetime extension, the `Thing` temporary would have been dropped before `x` is used. Could you explain why exactly does it compile?

```rust
#![feature(if_let_guard)]

fn main() {
    match () {
        _ if let x = &make_thing() => {
            let _a = x;
        },
        _ => {}
    }
}

struct Thing;

fn make_thing() -> Thing {
    Thing
}
```

ehuss commented on src/expressions/match-expr.md:
```diff
@@ -9,15 +9,35 @@ MatchExpression ->
     `}`
 
-Scrutinee -> Expression _except [StructExpression]_
+Scrutinee -> Expression _except_ [StructExpression]
```
```suggestion
Scrutinee -> Expression _except [StructExpression]_
```

dianne commented on src/destructors.md:
````diff
@@ -185,6 +182,43 @@ let declared_first = PrintOnDrop("Dropped last in outer scope");
+Local variables declared in a `match` expression or pattern-matching `match` guard are associated to the arm scope of the `match` arm that they are declared in.
+
+```rust
+# #![allow(irrefutable_let_patterns)]
````
Fixed: [(diff)](https://github.com/rust-lang/reference/compare/6f1597adebc92c4978a102753434212c39fdc34f..f1b193ce9e1f582125dbdddc9b42a2b569384f70#diff-76f8b59537d5af10271dfb21f24d9288b354500192420a6438ec9efbd119f986)

traviscross committed:
Swap order of `MatchConditions` alternatives

In our PEG-style grammar, the order matters, and in this case, we need
`MatchGuardChain` to come after `Expression`.

Thanks-to: Eric Huss <eric@huss.org>

ehuss reviewed (approved):
Thanks! Merging now that https://github.com/rust-lang/rust/pull/141295 has merged.
